chain/index/events.go

```diff
@@ -0,0 +1,560 @@
+	}
+
+	ems := make([]executedMessage, len(msgs))
+
+	for i := 0; i < len(msgs); i++ {
+		ems[i].msg = msgs[i]
+
+		var rct types.MessageReceipt
+		if found, err := receiptsArr.Get(uint64(i), &rct); err != nil {
+			return nil, xerrors.Errorf("failed to load receipt %d: %w", i, err)
+		} else if !found {
+			return nil, xerrors.Errorf("receipt %d not found", i)
+		}
+		ems[i].rct = rct
+
+		// no events in the receipt
+		if rct.EventsRoot == nil {
+			continue
+		}
+
+		eventsArr, err := amt4.LoadAMT(ctx, st, *rct.EventsRoot, amt4.UseTreeBitWidth(types.EventAMTBitwidth))
+		if err != nil {
+			if recomputeTipSetStateFunc == nil {
+				return nil, xerrors.Errorf("failed to load events amt for message %s: %w", ems[i].msg.Cid(), err)
+			}
+			log.Warnf("failed to load events amt for message %s: %s; recomputing tipset state to regenerate events", ems[i].msg.Cid(), err)
+
+			if err := recomputeTipSetStateFunc(ctx, msgTs); err != nil {
+				return nil, xerrors.Errorf("failed to recompute missing events; failed to recompute tipset state: %w", err)
+			}
+
+			eventsArr, err = amt4.LoadAMT(ctx, st, *rct.EventsRoot, amt4.UseTreeBitWidth(types.EventAMTBitwidth))
+			if err != nil {
+				return nil, xerrors.Errorf("failed to load events amt for re-executed tipset for message %s: %w", ems[i].msg.Cid(), err)
+			}
+
+			log.Infof("successfully recomputed tipset state and loaded events amt for message %s", ems[i].msg.Cid())
+		}
+
+		ems[i].evs = make([]types.Event, eventsArr.Len())
+		var evt types.Event
+		err = eventsArr.ForEach(ctx, func(u uint64, deferred *cbg.Deferred) error {
+			if u > math.MaxInt {
+				return xerrors.Errorf("too many events")
+			}
+			if err := evt.UnmarshalCBOR(bytes.NewReader(deferred.Raw)); err != nil {
+				return err
+			}
+
+			cpy := evt
+			ems[i].evs[int(u)] = cpy
+			return nil
+		})
+
+		if err != nil {
+			return nil, xerrors.Errorf("failed to iterate over events for message %d: %w", i, err)
+		}
+
+	}
+
+	return ems, nil
+}
+
+// checkTipsetIndexedStatus verifies if a specific tipset is indexed based on the EventFilter.
+// It returns nil if the tipset is indexed, ErrNotFound if it's not indexed or not specified,
+func (si *SqliteIndexer) checkTipsetIndexedStatus(ctx context.Context, f *EventFilter) error {
+	var tipsetKeyCid []byte
+	var err error
+
+	// Determine the tipset to check based on the filter
+	switch {
+	case f.TipsetCid != cid.Undef:
+		tipsetKeyCid = f.TipsetCid.Bytes()
+	case f.MinHeight >= 0 && f.MinHeight == f.MaxHeight:
+		tipsetKeyCid, err = si.getTipsetKeyCidByHeight(ctx, f.MinHeight)
+		if err != nil {
+			if err == ErrNotFound {
+				// this means that this is a null round and there exist no events for this epoch
+				return nil
+			}
+
+			return xerrors.Errorf("failed to get tipset key cid by height: %w", err)
+		}
+	default:
+		// Filter doesn't specify a specific tipset
+		return nil
```

Comment:
Are you sure? what about where `f.MinHeight >= 0 && f.MinHeight != f.MaxHeight`
What are we actually trying to achieve here? It's just a sanity check for "I got no events, even though I waited till the head was indexed". Then we're checking if we have data for what they requested? But we're only checking when they want a single tipset and leaving the range case as too-hard-to-check? Could we check min & max if they're set and not the same? Is it worth the trouble?

Reply:
@rvagg 

- This function is to help differentiate between "I have this tipset but not the events i.e. empty events is the correct response here" vs "I have not indexed/seen this tipset at all i.e. ErrNotFound"

- It solves a long standing UX problem with the Events RPC API where clients are not able to differentiate between missing events/valid empty set of events for a given tipset.

Reply:
This is harder and much more expensive to do for a range of tipsets.
Do we go and verify that each tipset in the range exists/each of them has events etc ?

This change will still be impactful for the **_most_** common use case of MinHeight == MaxHeight.

Reply:
yeah, maybe add a comment in here though, just so a future maintainer doesn't think this is an exhaustive check - we're just going for the straightforward cases here and there may be more complex ones that we're opting not to check

Reply:
Done.

Reply:
👌